proposed/MPID Messaging.md

```diff
@@ -0,0 +1,295 @@
+    mpid_header_name(mpid_message.signed_header)
+}
+
+pub fn sign_mpid_header(mpid_header: &MpidHeader, private_key: PrivateKey) -> Vec<u8> {
+    ::sodiumoxide::crypto::sign::sign(encode(mpid_header), private_key)
+}
+pub fn parse_signed_header(signed_header: &Vec<u8>, public_key: PublicKey) -> Result<MpidHeader, ::routing::error::Error> {
+    decode::<MpidHeader>(::sodiumoxide::crypto::sign::verify(signed_header, public_key))
+}
+```
+
+## Outbox
+
+This is a simple data structure for now and will be a hash map of serialised and encrypted `MpidMessage`s.  There will be one such map per MPID (owner), held on the MpidManagers, and synchronised by them at churn events.
+
+## Inbox
+
+This is an even simpler structure and again there will be one per MPID (owner), held on the MpidManagers, and synchronised by them at churn events.
+
+This can be implemented as a `Vec<(sender: ::routing::NameType, signed_header: Vec<u8>)>`.
+
+## Messaging Format Among Nodes
+
+Messages between Clients and MpidManagers will utilise [`::routing::structured_data::StructuredData`][5]:
+```rust
+let sd_for_mpid_message = StructuredData {
+    type_tag: MPID_MESSAGE_TAG,
+    identifier: mpid_message_name(mpid_message),
+    data: encode(mpid_message),
+    previous_owner_keys: vec![],
+    version: 0,
+    current_owner_keys: vec![mpid_message.sender],
+    previous_owner_signatures: vec![]
+}
+```
+```rust
+let sd_for_mpid_header = StructuredData {
+    type_tag: MPID_HEADER_TAG,
+    identifier: mpid_message_name(mpid_message),  // or mpid_header_name(signed_header)
+    data: mpid_message.signed_header,  // or inbox.signed_header
+    previous_owner_keys: vec![],
+    version: 0,
+    current_owner_keys: vec![mpid_message.sender],  // or vec![inbox.sender]
+    previous_owner_signatures: vec![]
+}
+```
+
+## Message Flow
+
+```
+                    MpidManagers(A)    MpidManagers(B)
+                   /      *                   *        \
+MpidClient(A) ->  -       *                   *         -  <- MpidClient(B)
+                   \      *                   *        /
+
+```
+1. `MpidClient(A)` sends an `MpidMessage` to `MpidManagers(A)` as a Put request.
+1. `MpidManagers(A)` store this message in their outbox for A and send the `signed_header` component to `MpidManagers(B)` as a Put request, unless the outbox is full in which case a PutFailure response is returned to `MpidClient(A)`.
+1. `MpidManagers(B)` try to store the `(sender, signed_header)` in their inbox for B.  If successful, they forward it to `MpidClient(B)` as a Put request immediately or as soon as it appears online.  If unsuccessful (e.g. the inbox is full or sender has been blacklisted), they reply to `MpidManagers(A)` with a PutFailure, who then remove the message from the outbox and send a PutFailure response to `MpidClient(A)`.
+1. To retrieve the message from the network, `MpidClient(B)` sends a Get request to `MpidManagers(A)`.
```

Comment:
A Few questions here:
1. If the steps are mandatory for the clients to perform, it is better to do it at the Vault level to make it more deterministic, i.e. when MpidManagers(A) receive the `GET` request from `Client(B)`, they should do both the steps - give data to the client **and** initiate purge stuff from MpidManagers(B) and self. That way it is closer to being Network-atomic than if the clients are involved. This is consequent to the notion that Clients are temporary entities on the Network while Vaults are statistically permanent thus operations initiated by them are more deterministic.
2. Inbox never seems to get the actual content. If there is a need for data to be stored in Inbox even after the Client has retrieved it, then with this approach the sender will have to sacrifice his storage (since it is only there that data is maintained). Could it be more useful if `B` could retrieve messages from `MpidManagers(A)` into `MpidManagers(B)` (i.e. move them) during `GET`s so that the messages could be stored there till `B` wants them deleted ?